File: Sequence/q1-q99/q80-q89/q81/q81.cpp
class Solution {
public:
    bool search(vector<int>& nums, int target) {
        int left = 0, right = nums.size() - 1, middle = 0;
        while (left <= right) {
            middle = (right + left) / 2;
            if (nums[middle] == target || nums[left] == target || nums[right] == target) {
                return true;
            }
            if (nums[middle] < nums[right]) {
                if (nums[middle] < target && target < nums[right]) {
                    left = middle + 1;
                } else {
                    right = middle - 1;
                }
            } else if (nums[middle] > nums[left]) {
                if (nums[middle] > target && nums[left] < target) {
                    right = middle - 1;
                } else {
                    left = middle + 1;
                }
            } else {
                right--;
            }
        }
        return false;
            
    }
};

File: Sequence/q1-q99/q80-q89/q81/q81.py
class Solution:
    def search(self, nums: List[int], target: int) -> bool:
        left, right, middle = 0, len(nums) - 1, 0
        while left <= right:
            middle = (left + right) // 2
            if nums[left] == target or nums[right] == target or nums[middle] == target:
                return True
            if nums[middle] < nums[right]:
                if nums[middle] < target and target < nums[right]:
                    left = middle + 1
                else:
                    right = middle - 1
            elif nums[middle] > nums[left]:
                if nums[middle] > target and nums[left] < target:
                    right = middle - 1
                else:
                    left = middle + 1
            else:
                right -= 1
        return False
